MudTextField: Check if user is trying to enter negative number

Lewis-Pitman opened the pull request:
## Description
Check if user is trying to enter negative number. If they are, modify the text so that it allows them to do so.
This seems like a hacky fix. May need guidance on an alternative fix
For issue #11055 

## How Has This Been Tested?
Visually

## Type of Changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

Before:
https://github.com/user-attachments/assets/be2f42db-a9c4-4f5d-93bd-d68fb7f18289

After:
https://github.com/user-attachments/assets/21621dd6-ba74-4e99-b4f8-c2b0056671c1

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [ ] I've added relevant tests.

Lewis-Pitman committed:
Check if user is trying to enter negative number

Check if user is trying to enter negative number. If they are, modify the text so that it allows them to do so.

vernou commented on src/MudBlazor/Base/MudBaseInput.cs:
```diff
@@ -437,7 +437,21 @@ protected virtual async Task SetTextAsync(string? text, bool updateValue = true)
-                Text = text;
+                if (text == "0-")
+                {
+                    Text = "-";
```
If `MudTextField` is bound to `string`, `0-` can be a valid input.

vernou reviewed (changes requested):
Thank for your help and your diagnostic.

`MudBaseInput` is used by many types, so every change is risky.
So maybe you try to fix elsewhere.

Also, can you add some test?

Lewis-Pitman commented:
This is a change in MudTextField.razor.cs that gives the desired effect. I have not yet implemented this for MudNumericField. Does this look better? @vernou 
![image](https://github.com/user-attachments/assets/3cfce5f1-8546-4030-9b63-8b42ca0c30c2)

Nickztar commented:
Would probably have to check for more types of Value, or maybe it would be possible to check if Value is [`INumber` ](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.inumber-1?view=net-9.0)

Lewis-Pitman committed:
MudTextField allow negative input if numeric

Allow the user to type negative numbers correctly if bound to a numeric type

Lewis-Pitman commented:
This code makes MudTextField work as expected. Checking if the value was INumber did not work. Happy to implement tests if this method is accepted.

As for MudNumericField, I have been debugging the issue for hours with no luck. I think modifying MudBaseInput may be necessary to get it to work. Using SetTextAsync on MudNumericField doesn't change anything. I even tried intercepting the - key inside HandleKeyDownAsync but had no luck getting the desired outcome. I am unsure what to do here

Lewis-Pitman commented:
> As for MudNumericField, I have been debugging the issue for hours with no luck.

Seems like there is a seperate issue #11035 for MudNumericField specifically. Since the solution seems to be a bit more complex, it might be better for me to leave it for now and someone else can try this?

Lewis-Pitman commented:
Also just noticed that I'll need to add billable types too like int? And float? For example

Lewis-Pitman commented:
Also just noticed that I'll need to add nullable types too like int? And float? For example

Lewis-Pitman committed:
MudTextField: Negative number on first attempt

- MudTextField allows a negative number to be entered on the first attempt when bound to a numeric type
- Added test to check this works
- Reverted MudBaseInput changes

vernou commented on src/MudBlazor.UnitTests/Components/TextFieldTests.cs:
```diff
@@ -1578,6 +1578,20 @@ public void Should_render_conversion_error_message()
+                .Add(p => p.Immediate, true)
+                .Add(p => p.Value, 5));
+
+            comp.Find("input").Input("0-");
```
Before, can you simulate the user that remove all char in input?
Something like :
```csharp
comp.Find("input").Input("");
comp.Instance.Text.Should().Be("0");
comp.Instance.Value.Should().Be(0);
```

vernou commented on src/MudBlazor.UnitTests/Components/TextFieldTests.cs:
```diff
@@ -1578,6 +1578,20 @@ public void Should_render_conversion_error_message()
+            comp.Find("input").Input("0-");
+
+            comp.Instance.Text.Should().Be("-");
+            comp.Instance.Text.Should().NotBe("0");
```
If `Text` is `"-"`, then `Text` will never be isn't `"0"`.
So this verification is not necessary.

Lewis-Pitman committed:
Revert MudBaseInput and improve test

Reverted MudBaseInput to what it was before. This was not properly reverted before, causing a test to run infinitely

Improved the test by checking if entering nothing changes the value of 5 to 0.

Lewis-Pitman commented:
I realised that I hadn't reverted MudBaseInput properly after the initial changes I made to it, which caused one of the checks to fail. I have now properly reverted the code to how it was before and this has been fixed.

As for the test, entering "" causes `comp.Instance.Text.Should().Be("0");` to fail, although `comp.Instance.Value.Should().Be(0);` does pass meaning it changed the value from 5 to 0 when entering nothing. I removed the text check, but please let me know if I should add it back in.

vernou commented:
It's important to check `Text` and `Value` for each step. Can you reput it? Just adapt the expected.

Lewis-Pitman committed:
TextFieldTests negative number improvement

vernou commented on src/MudBlazor.UnitTests/Components/TextFieldTests.cs:
```diff
@@ -1578,6 +1578,28 @@ public void Should_render_conversion_error_message()
+                .Add(p => p.Value, 5));
+
+            comp.Find("input").Input("");
+            comp.Find("input").Blur();
```
With `Immediate`, the blur isn't necessary.

Lewis-Pitman committed:
Remove blur from test

Remove blur from test

vernou commented on src/MudBlazor.UnitTests/Components/TextFieldTests.cs:
```diff
@@ -1578,6 +1578,23 @@ public void Should_render_conversion_error_message()
+            comp.Find("input").Input("5");
+            comp.Find("input").Input("0-");
+            comp.Instance.Text.Should().Be("-");
+            comp.Instance.Value.Should().Be(5);
```
Can you add a step `comp.Find("input").Input("-7");` to check the full scenario?

Lewis-Pitman committed:
Test full scenario

Added test to check entering a negative number
